EventPhases/explainer.md

```diff
@@ -0,0 +1,501 @@
+
+> Incidentally, this is why CSS-in-JS is such a terrible antipattern. Manipulating style rules at runtime is dramatically more costly than poorly timed style readback because it also blows away the caches that make style recalculation and layout faster.
+
+Most code that manipulates the DOM, then calls any API that returns sizing information _before_ yielding to the main thread causes style recalculation and layout to run, which can be extremely slow.
+
+```html
+<script type="module">
+  // Assuming no script before this code modifies the DOM or styles,
+  // this code will execute quickly:
+  let foo = document.querySelector("#foo");
+  // Synchronously reads back dimensions computed at last layout
+  let width = foo.innerWidth; 
+  // Set a style that will impact the element's computed width the
+  // next time a layout occurs.
+  foo.style.width = `${parseInt(width) + 10}px`;
+  // ...
+  foo.style.maxWidth = "75%";
+  // Explicitly yield to main thread here.
+  scheduler.yield();
+  // Because the DOM is "dirty" (has chaged since JS started 
+  // running), all frame-generation phases will execute:
+  // style recalc -> layout -> paint -> composite.
+</script>
+```
+
+It's straightforward to apply the "read, then write" discipline in small programs, but imagine an only slightly more complicated program that uses code from a different vendor; a common occurrence in the modern web:
+
+```js
+// analytics.example.com/analytics.js
+export function log(element) {
+  // ...
+  let height = element.innerHeight;
+  // ...
+}
+```
+
+```html
+<script type="module">
+  import * from "https://analytics.example.com/analytics.js"
+
+  let foo = document.querySelector("#foo");
+  let width = foo.innerWidth; 
+  foo.style.width = `${parseInt(width) + 10}px`;
+  log(foo);
+  // ... 
+  foo.style.maxWidth = "75%";
+  // Yield the thread.
+  scheduler.yield();
+  // style recalc -> layout -> paint -> composite.
+</script>
+```
+
+In this (only slightly) stylized example, `log()` performs a synchronous readback of style information that, implies that *at least* the style recalculation and layout phases of the frame generation pipeline must be re-run in order to compute the element's `innerHeight`. This happens at when the `innerHeight` accessor is called because browsers attempt to batch writes in the hopes that they will not need to perform the expensive style, layout, paint, and composite phases until a full frame needs to be generated for the user. 
```

Comment:
```suggestion
In this (only slightly) stylized example, `log()` performs a synchronous readback of style information that implies that *at least* the style recalculation and layout phases of the frame generation pipeline must be re-run in order to compute the element's `innerHeight`. This happens at when the `innerHeight` accessor is called because browsers attempt to batch writes in the hopes that they will not need to perform the expensive style, layout, paint, and composite phases until a full frame needs to be generated for the user. 
```

Reply:
oh, wait, I see it now. Thanks for the help!